### ISSUE TITLE
Should `deps` be built in the base image? 

### ISSUE THREAD

**Issue by ethanknox-palmetto** [palmetto/palm-dbt#14]
### what is the general idea or problem you want to solve?
currently the stock base image created by `palm init` does not contain deps. most of the commands like `run`, `test`, `cycle` etc will execute deps first, so you always have fresh deps (though this does slow them down significantly). If we move deps to the build cycle, what are the pros? does this make sense? 

### Desired Implementation
run `dbt deps` in the image build. Then move deps to a stand-alone palm command and a flag for the other commands. 

### Risks
this could leave you working with outdated deps, and when a new container is built the behavior might be unexpected. Especially if dbt packages are poorly pinned.

**Comment by ethanknox-palmetto** [comment]
@mrogers-palmetto do you have reservations around this at all? I think the good probably outweighs the bad IMO

**Comment by mrogers-palmetto** [comment]
I agree with having deps run separately, either in the image or (just thought of!) an init step!

We already have `palm init` to set up the local palm environment. What if we add an `init` command to plugins that will get run as part of `palm init` after the core palm init? This could make palm act similarly to Terraform, where you re-init when you make a big change, like importing a new package into `dbt`, but for most use cases you just have to init once at the beginning of your workflow.

Unfortunately, the current strategy of complete override behavior of commands from core -> plugin -> repo means that adding an init command to `palm-dbt` will only mess up palm core's init... But this could be a special case command where init cascades rather than overwrites.

I think `deps` generally _should_ be decoupled from the primary dbt commands, though. It doesn't need to happen repeatedly. In the Dockerfile seems as good a place as any in the absence of complete `init` functionality.

**Comment by ethanknox-palmetto** [comment]
I'm inclined to use a new command like 'deps' so we can keep init nice and tidy; actually, if we are volume-mounting the project into the image when the image has deps built in, you get these nice transient deps that are inspectable but never in the actual codebase. 

we could also make the other commands smarter across palm-dbt so they always check for deps and run deps if none exist. that could be a `base_dbt` command that takes the context and execution string, or something like that?

**Comment by jakeberesford-palmetto** [comment]
One thing we want to avoid with any implementation is requiring the user to remember they have to run deps. I've lost count of the number of times I've run in to that rather confusing error you get from DBT when deps aren't installed. Even having to run it manually one time is bad, because I will forget.

I like putting it in the docker image for this reason.

Locally, I see no issue with moving deps into the Dockerfile. The issue, as you said, would be the risk of stale deps in production if a given image hasn't been rebuilt in some time.

I'm curious how other containerized dbt projects are handling this. It'd be good to implement a solution here that meets the existing expectations of the dbt community.

**Comment by mrogers-palmetto** [comment]
Worth updating here that we ran into a large hurdle 😓  because running `deps` in the image build step gets clobbered by the runtime volume mount. Thus, a workaround is required, building `deps` into a different path than the path for the volume, and changing the dbt project configuration for `modules-path` (`packages-install-path` in dbt v1.0.0+) to point at that new destination.

**Comment by jakeberesford-palmetto** [comment]
So, as I'm working through this change to doing `dbt deps` in the Dockerfile, and the adjacent changes (putting dbt_modules in a different location) I'm realizing that this may not be a good 'default' to share/enforce in palm-dbt.

We need to think about how we want to proceed with this change. Here are some of the options I'm weighing, they kind of build on each-other:

1. Do not make this change, run deps on every run as we have been doing.
2. Make the change, tell other people they need to set their `modules-path` to be outside of the `/app` directory, consider setting that for them during containerize?
3. Make the change and add a `--no-deps` option to all relevant commands, we have to remember to pass that option each time
4. Make the change, introduce `plugin_config` which we couple with the `--no-deps` logic so we dont have to remember that flag.

Worth noting that plugin config is something we've talked about at various times, but so far haven't really needed. I'm also not sure what the implementation of that would look like. The logical place for the config would be on the plugin instance. However, commands are decoupled from the Plugin instance, so the config would also need to be decoupled (sad). One approach would be adding a `plugin_config` method to the palm environment, so the implementation in the command would look something like this:

```
@click.option("--no-deps", is_flag=True, help="will skip clean & deps")
@click.pass_obj
def cli(environment, no_deps: bool)
    no_deps = environment.plugin_config('dbt').get('skip-deps', no_deps)
```
 
This is obviously not a small change, and I'm reluctant to make a hasty decision that we might regret later.

Curious to hear what you all think about this?

**Comment by ethanknox-palmetto** [comment]
OK how about this. Breaking change but still relatively small.
- start with a warning "Starting in palm-dbt v0.3.0 deps will not be reloaded by default. Use --deps flag to force a refresh of deps") 
- look for deps, if they exist echo ("deps found, using existing deps.")
- if deps are not found, run dbt deps.

no dockerfile gnarling, everything works as expected. 
still bake the deps into the dockerfile so it will not run deps in production :)

**Comment by emekdahl-palmetto** [comment]
@jakeberesford-palmetto I like the idea of a spot in the config where people can set whether they want to run deps by default or not, combined with the ability to modify that by flag.

**Comment by jakeberesford-palmetto** [comment]
> OK how about this. Breaking change but still relatively small.
> 
> * start with a warning "Starting in palm-dbt v0.3.0 deps will not be reloaded by default. Use --deps flag to force a refresh of deps")
> * look for deps, if they exist echo ("deps found, using existing deps.")
> * if deps are not found, run dbt deps.
> 
> no dockerfile gnarling, everything works as expected. still bake the deps into the dockerfile so it will not run deps in production :)

Oh yeah, that was the other option... explicitly opting into deps on a given command. I got carried away by my train of thought and forgot to write that one down!

**Comment by jakeberesford-palmetto** [comment]
> @jakeberesford-palmetto I like the idea of a spot in the config where people can set whether they want to run deps by default or not, combined with the ability to modify that by flag.

I agree that regardless of whether we are saying opt-in to deps with `--deps` or opt out of deps with `--no-deps` it'd be a better DX to have a project-level configuration along the lines of my comment. That said, I would also like to take a little more time to think about how plugin_config should work. If we pick a sensible path forward for now, we can add in configurable defaults later — once we've decided on the pattern for plugin configuration.

I started a discussion here so that we can share our thoughts on plugin_config as a group: https://github.com/palmetto/palm-cli/discussions/54

**Comment by ethanknox-palmetto** [comment]
> "looking for deps" could get kind of messy... we'd have to read the module-path config from the dbt_project, if it's not present, check for the packages_install_path (renamed in dbt v1.0) and then if neither are present, fall back to the default ./dbt_modules directory.

I feel like this would be OK, it's actually not terrible (a little yaml parsing); I would also be willing to bet that nearly all projects use the default location for deps. _and_ if we are wrong and we don't detect deps where there is deps, it doesn't hurt anything - it is just a little slower because it runs deps every time. From a "most good with least pain" standpoint that feels like the right answer to me. Look for the default location, if it's there great if not run deps. Makes life better for 90% of situations and doesn't hurt the edge cases, just helps them less

**Comment by jakeberesford-palmetto** [comment]
After all this talk, and thought, and toil... I think we have a better way! By fixing the volume mount issue directly instead of doing insane things to try to wiggle around it, we don't need to worry too much about that config. Instead, here's how I think it will work:

1. Our convention is: deps are installed in dockerfile
2. If you need to update your deps, you rebuild the dockerfile with `palm build`

That is it. 
No flags, because that just makes a mess and won't persist between runs.
No `palm deps` because that also won't persist between runs.
If you ran `dbt clean` or `palm cleanup` and blew away your deps, guess what? Doesn't matter... they are in the image and they will come right back the next time you run a container.

There may be a future where we need to add a `--deps` flag to a bunch of commands because people don't want to build deps into their images for some reason. I'd be interested to hear their reasons, and if they are valid we could definitely add that - preferably as a config, rather than an option that they need to remember!

**Comment by mrogers-palmetto** [comment]
Nice. I like it. It does still enforce a convention which is a bit more convoluted to understand than the standard `dbt` practice for deps, but the simplicity of it for users makes it seem more than worth it. We should also maybe identify deps exceptions and do a smart warning to users to re-run `palm build`, because deps error messages are not very obvious (just complains that a macro isn't defined or something).

**Comment by mrogers-palmetto** [comment]
BUT, what happens when the user overrides their modules/packages path?

**Comment by jakeberesford-palmetto** [comment]
> Nice. I like it. It does still enforce a convention which is a bit more convoluted to understand than the standard dbt practice for deps, but the simplicity of it for users makes it seem more than worth it. We should also maybe identify deps exceptions and do a smart warning to users to re-run palm build, because deps error messages are not very obvious (just complains that a macro isn't defined or something).

We could add some error detection in the future, if this starts to become a pain point.

> BUT, what happens when the user overrides their modules/packages path?

I added some functionality to `palm containerize` that will detect v1.0 and take care of the configs for both modules/packages path. If they change it after the fact, they would have to update the volume mount, too.

**Comment by mariahjrogers** [comment]
Resolved by #44
